karapace/schema_registry.py

```diff
@@ -145,130 +143,131 @@ async def get_master(self, ignore_readiness: bool = False) -> Tuple[bool, Option
                     return are_we_master, master_url
                 await asyncio.sleep(1.0)
 
-    def get_compatibility_mode(self, subject: SubjectData) -> CompatibilityModes:
-        compatibility = subject.get("compatibility", self.config["compatibility"])
+    def get_compatibility_mode(self, subject: Subject) -> CompatibilityModes:
+        compatibility = self.database.get_subject_compatibility(subject=subject)
+        if compatibility is None:
+            # If no subject compatiblity found, use global compatibility
+            compatibility = cast(str, self.config["compatibility"])
         try:
             compatibility_mode = CompatibilityModes(compatibility)
         except ValueError as e:
             raise ValueError(f"Unknown compatibility mode {compatibility}") from e
         return compatibility_mode
 
-    async def schemas_list(self, *, include_deleted: bool, latest_only: bool) -> Dict[Subject, SubjectData]:
+    async def schemas_list(self, *, include_deleted: bool, latest_only: bool) -> Dict[Subject, List[SchemaVersion]]:
         async with self.schema_lock:
-            return self.schema_reader.get_schemas_list(include_deleted=include_deleted, latest_only=latest_only)
+            schemas = self.database.find_schemas(include_deleted=include_deleted, latest_only=latest_only)
+            return schemas
 
     def schemas_get(self, schema_id: int, *, fetch_max_id: bool = False) -> Optional[TypedSchema]:
-        with self.schema_reader.id_lock:
-            try:
-                schema = self.schema_reader.schemas.get(schema_id)
+        try:
+            schema = self.database.find_schema(schema_id=schema_id)
 
-                if schema and fetch_max_id:
-                    schema.max_id = self.schema_reader.global_schema_id
+            if schema and fetch_max_id:
+                schema.max_id = self.database.global_schema_id
 
-                return schema
-            except KeyError:
-                return None
+            return schema
+        except KeyError:
+            return None
```

Comment:
issue: My usual pet peeve, I think we should have only `schema = ...` within the try-block here.